PythonVisualizations/PriorityQueue.py

```diff
@@ -0,0 +1,367 @@
+        self.nItems = 0
+        self.buttons = self.makeButtons()
+        self.index = self.createIndex(self.nItems-1, 'front', level=-1) # indicate priority
+        
+        self.display()
+
+    def __str__(self):
+        return str(self.list)
+
+    # ARRAY FUNCTIONALITY
+
+    def createIndex(self, index,  # cell
+                    name=None,  # with an optional name label
+                    level=1,  # at a particular level away from the cells
+                    color=None):  # (negative are below)
+        if not color: color = self.VARIABLE_COLOR
+
+        cell_coords = self.cellCoords(index)
+        cell_center = self.cellCenter(index)
+        level_spacing = self.VARIABLE_FONT[1]
+        x = cell_center[0]
+        if level > 0:
+            y0 = cell_coords[1] - self.CELL_SIZE * 3 // 5 - level * level_spacing
+            y1 = cell_coords[1] - self.CELL_SIZE * 3 // 10
+        else:
+            y0 = cell_coords[3] + self.CELL_SIZE * 3 // 5 - level * level_spacing
+            y1 = cell_coords[3] + self.CELL_SIZE * 3 // 10
+        arrow = self.canvas.create_line(
+            x, y0, x, y1, arrow="last", fill=color)
+        if name:
+            label = self.canvas.create_text(
+                x + 2, y0, text=name, anchor=SW if level > 0 else NW,
+                font=self.VARIABLE_FONT, fill=color)
+        return (arrow, label) if name else (arrow,)
+
+    def insert(self, val):
+        callEnviron = self.createCallEnvironment()
+        self.startAnimations()
+        j = self.nItems - 1  # Start at front
+
+        indexJ = self.createIndex(j, 'j', level=-2)
+        callEnviron |= set(indexJ)
+
+        #self.list[j] = drawable(None)
+        self.list[self.nItems] = drawable(None)
+        
+        while j >= 0 and val >= self.list[j].val:  # Move bigger items right
+
+            self.moveItemsBy(indexJ, (-self.CELL_SIZE, 0), sleepTime=0.1)  # Move "j" arrow
+            self.wait(1)
+            
+            self.assignElement(j+1, j, callEnviron)
+            self.list[j+1].val = self.list[j].val
+            j -= 1
+            self.wait(1)
+
+        self.nItems += 1
+            
+        # Location of the new cell in the array
+        toPositions = (self.cellCoords(j+1),
+                       self.cellCenter(j+1))
+
+        # Animate arrival of new value from operations panel area
+        canvasDimensions = self.widgetDimensions(self.canvas)
+        startPosition = [canvasDimensions[0] // 2, canvasDimensions[1]] * 2
+        startPosition = add_vector(startPosition, (0, 0, self.CELL_SIZE, self.CELL_SIZE))
+        cellPair = self.createCellValue(startPosition, val)
+        callEnviron |= set(cellPair)  # Mark new cell as temporary
+        self.moveItemsTo(cellPair, toPositions, steps=self.CELL_SIZE, sleepTime=0.01)
+        callEnviron -= set(cellPair)  # New cell is no longer temporary
+        self.list[j] = (drawable(
+            val, self.canvas.itemconfigure(cellPair[0], 'fill')[-1]))
```

Comment:
In this section the new item in `cellPair` is moved into place, then assigned to the array.  Taking the lead from the textbook code, lines 94-97 should be replaced with:
```
        self.moveItemsTo(cellPair, toPositions, steps=self.CELL_SIZE, sleepTime=0.01)
        self.list[j+1] = drawable(
            val, self.canvas.itemconfigure(cellPair[0], 'fill')[-1], *cellPair)
        callEnviron -= set(cellPair)  # New cell is no longer temporary

        self.nItems += 1
        self.moveItemsBy(self.index, (self.CELL_SIZE, 0), sleepTime=0.01)
```

Here's why:
* The assignment is made right after the `cellPair` is in position.
* The assignment goes to cell `j+1` in the textbook code, and should do the same here.
* The assignment was only creating a `drawable` with a `val` and `color` field.  Adding the `*cellPair` to the end of the argument list, passes those two canvas items to the `drawable` constructor to put in `display_shape` and `display_val`.
* The `callEnviron` should only be updated after the `cellPair` is assigned to `self.list`.  That's what makes it a lasting item in the data structure.
* Now that the assignment is done, `nItems` can be incremented.
* Since the `nItems` value changed, we move the index pointer on the canvas.